src/vos/evtree.c

```diff
@@ -1172,20 +1133,52 @@ evt_node_mbr_get(struct evt_context *tcx, struct evt_node *node)
 	return &node->tn_mbr;
 }
 
+int
+evt_dtx_check_visibility(struct evt_context *tcx, struct evt_desc *desc,
+			 uint32_t intent)
+{
+	return vos_dtx_check_visibility(evt_umm(tcx), tcx->tc_coh,
+					umem_ptr2id(evt_umm(tcx), desc), intent,
+					DTX_RT_EVT);
+}
+
 /** (Re)compute MBR for a tree node */
 static void
 evt_node_mbr_cal(struct evt_context *tcx, struct evt_node *node)
 {
 	struct evt_rect *mbr;
 	int		 i;
+	bool		 leaf;
 
 	D_ASSERT(node->tn_nr != 0);
 
 	mbr = &node->tn_mbr;
 	*mbr = *evt_node_rect_at(tcx, node, 0);
+	leaf = evt_node_is_leaf(tcx, node);
 	for (i = 1; i < node->tn_nr; i++) {
 		struct evt_rect *rect;
 
+		if (leaf) {
+			struct evt_desc	*desc;
+			int		 rc;
+
+			desc = evt_node_desc_at(tcx, node, i);
+			rc = evt_dtx_check_visibility(tcx, desc,
+						      DAOS_INTENT_DEFAULT);
+			/* Skip the invisible record. */
+			if (rc == DTX_VBT_INVISIBLE)
+				continue;
+
+			/* DER_INPROGRESS means that we are not sure about the
+			 * record visibility, then count it. If related record
+			 * is aborted later, then it can be excluded next time.
+			 */
+			if (rc < 0 && rc != -DER_INPROGRESS) {
+				D_ERROR("Unknown record visiblity, skip it\n");
+				continue;
+			}
+		}
+
```

Comment:
As per our discussion, I don't think excluding aborted rectangle from MBR would buy us any obvious value, on the contrary, it introduces complexity (for example, we need to handle empty MBR case).

Given that the 'visibility' (2PC state, actually) isn't involved in indexing by design, and we didn't exclude aborted key hash from interior nodes for btree, I suggest we don't do it for evtree as well.

Reply:
I will remove the visibility check from the MBR handling logic, that may affect the subsequent lookup efficiency, but I think it will not affect the correctness. Let's see what will happen, if need, we can add them back in the future.